### PR TITLE
libnmstate: Interface desire state sanitize adjustment

### PR DESCRIPTION
**Pull request by EdDev**
nmstate is processing the desired state of an interface and adds base
information for the 'ethernet' properties (auto-negotiation, speed and
duplex).
The 'ethernet' properties are relevant only for ethernet type interfaces
and not for virtual ones. In addition, the base values used in the
sanitize step may cause the post apply verification to fail.

The interface ethernet state sanitize is now limited to ethernet type
interfaces and special handling has been added in the verification step
to allow proper comparison between the desired and the current state.

### CODE REVIEW

File: libnmstate/netapplier.py [hunk 1]
```diff
@@ -58,7 +59,7 @@ def _apply_ifaces_state(interfaces_desired_state, verify_change):
             _edit_interfaces(ifaces_desired_state, ifaces_current_state)
         if verify_change:
             # FIXME: Remove this sleep after adding wait for state mechanism.
-            time.sleep(2)
+            time.sleep(5)
```
Comment (tyll) [inline libnmstate/netapplier.py]: Is this intentional here? Maybe this should be its own commit with an explanation.

Comment (EdDev) [reply libnmstate/netapplier.py]: No, a leftover. Thanks.

File: libnmstate/netapplier.py [hunk 2]
```diff
@@ -120,19 +121,22 @@ def generate_ifaces_metadata(ifaces_desired_state, ifaces_current_state):
...
     """
-    for iface_state in six.viewvalues(ifaces_desired_state):
-        ethernet = iface_state.setdefault("ethernet", {})
-        ethernet.setdefault("auto-negotiation", None)
-        ethernet.setdefault("speed", None)
-        ethernet.setdefault("duplex", None)
+    for ifname, iface_state in six.viewitems(ifaces_desired_state):
+        if ifname in ifaces_current_state:
+            iface_current_state = ifaces_current_state[ifname]
+            if iface_current_state['type'] == 'ethernet':
```
Comment (tyll) [inline libnmstate/netapplier.py]: IMHO it would be better to check if the current state contains an ethernet subtree here. The method to create the current state needs to be changed to only add the ethernet subtree if the interface type is ethernet.

Also wired.create_setting() should only be called if the interface type is ethernet.

File: libnmstate/netapplier.py [hunk 3]
```diff
@@ -120,19 +121,22 @@ def generate_ifaces_metadata(ifaces_desired_state, ifaces_current_state):
     )
 
 
-def sanitize_ethernet_state(ifaces_desired_state):
+def sanitize_ethernet_state(ifaces_desired_state, ifaces_current_state):
     """
     If auto-negotiation, speed and duplex settings are not provided, they need
```
Comment (tyll) [inline libnmstate/netapplier.py]:
```suggestion
    If auto-negotiation, speed and duplex settings are not provided, but exist in the current state, they need
```

File: libnmstate/netapplier.py [hunk 4]
```diff
@@ -120,19 +121,22 @@ def generate_ifaces_metadata(ifaces_desired_state, ifaces_current_state):
...
     from the actual configuration.  This makes it possible to distiguish
     whether a user specified these values in the later configuration step.
     """
-    for iface_state in six.viewvalues(ifaces_desired_state):
-        ethernet = iface_state.setdefault("ethernet", {})
-        ethernet.setdefault("auto-negotiation", None)
-        ethernet.setdefault("speed", None)
-        ethernet.setdefault("duplex", None)
+    for ifname, iface_state in six.viewitems(ifaces_desired_state):
+        if ifname in ifaces_current_state:
```
Comment (tyll) [inline libnmstate/netapplier.py]: AFAICS this could be simplified to something like

```
if 'ethernet' in ifaces_current_state.get(ifname, {})
    ethernet = ...
```

File: libnmstate/netapplier.py [hunk 5]
```diff
@@ -306,6 +310,15 @@ def assert_ifaces_state(ifaces_desired_state, ifaces_current_state):
         iface_cstate = ifaces_current_state[ifname]
         iface_dstate = _canonicalize_desired_state(
             ifaces_desired_state[ifname], iface_cstate)
+
+        iface_ethernet_dstate = iface_dstate.get('ethernet')
```
Comment (tyll) [inline libnmstate/netapplier.py]: IMHO it should be explained what is going on here to be able to know this still in two weeks. I needed the comment in the sanitize method as well to be able to understand the issue. :-)
Maybe it would be also better to put in its own method to be able to use variables such as current_state and desired_state since `...cstate/dstate` are hard to distinguish.

File: libnmstate/netapplier.py (hunk 3 above)
Comment (EdDev) [reply libnmstate/netapplier.py]: Done

File: libnmstate/netapplier.py (hunk 4 above)
Comment (EdDev) [reply libnmstate/netapplier.py]: I massaged it a bit.
I think we should check against the type and not against a subsequent property.

File: libnmstate/netapplier.py (hunk 2 above)
Comment (EdDev) [reply libnmstate/netapplier.py]: I will add the changes not to report the `ethernet` subtree, but see my previous response regarding the verification here.
Thanks for the two points.

File: libnmstate/netapplier.py (hunk 5 above)
Comment (EdDev) [reply libnmstate/netapplier.py]: Extracting to smaller methods is always good, thanks for pointing this out.

File: libnmstate/netapplier.py [hunk 6]
```diff
@@ -120,19 +121,22 @@ def generate_ifaces_metadata(ifaces_desired_state, ifaces_current_state):
...
 
 
-def sanitize_ethernet_state(ifaces_desired_state):
+def sanitize_ethernet_state(ifaces_desired_state, ifaces_current_state):
     """
-    If auto-negotiation, speed and duplex settings are not provided, they need
-    to be set to None to not override them with the values from the current
-    settings since the current settings are read from the device state and not
+    If auto-negotiation, speed and duplex settings are not provided,
+    but exist in the current statet, they need to be set to None
```
Comment (tyll) [inline libnmstate/netapplier.py]: There is a typo: `statet`

**Comment by tyll** [comment]
There is a typo in the commit message (desire vs desired) and I think it should start with a verb after the colon and include the word `ethernet`. Maybe it should even be separate commits

- ignore unspecified ethernet settings for verification
- sanitize ethernet settings only for type ethernet
- report ethernet settings only for type ethernet (this change is currently missing)
- change wired settings only for type ethernet (this change is currently misssing)

File: libnmstate/netapplier.py (hunk 6 above)
Comment (EdDev) [reply libnmstate/netapplier.py]: Done

**Comment by EdDev** [comment]
> There is a typo in the commit message (desire vs desired) and I think it should start with a verb after the colon and include the word ethernet. Maybe it should even be separate commits
> 
Ok, fair enough.
>     ignore unspecified ethernet settings for verification
>     sanitize ethernet settings only for type ethernet
>     report ethernet settings only for type ethernet (this change is currently missing)
Adding
>     change wired settings only for type ethernet (this change is currently misssing)
I answered on this but github hides the comment after changing the line.
"For the wired setting, it will not work as there are properties that are used there for virtual devices (like mtu)."

**Comment by EdDev** [comment]
ci test please
